File: ACL.java
//Advanced Computer Language interpreter

import java.util.Scanner;
import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;

public class ACL {
  
  //reads source file and returns the code
  static String getCode() {
    try {
      Scanner getName = new Scanner(System.in);
      System.out.println("File to interpret?");
      String fileName = getName.nextLine();
      getName.close();
    
      File file = new File(fileName);
      Scanner fRead = new Scanner(file);
      String code = fRead.nextLine();
      fRead.close();
    
      return code;
    } catch (FileNotFoundException e) {
      System.out.println("FILE FAIL");
      return "F";
    }
  }
 
  //binary to decimal
  static int toDec(String bin) {
    int l = bin.length();
    int j = l-1;
    int x = 0;
    int dec = 0;
    while (x < l) {
      dec += Math.pow(2, x) * ((int) bin.charAt(j) - 48);
      x++; j--;
    }
    return dec;
  }
  
  //--------------------------------------------------------------
  
  public static void main(String[] args) {
  
    //declares necessary variables etc.
    
    String code = getCode();
    int l = code.length();
    int i = 0; //code reader pointer; 
    
    ArrayList<Byte> memory = new ArrayList<Byte>(); //cell based, binary
    memory.add((byte) 0); //make the memory have one cell at start
    int ptr = 0; //memory pointer
    int cells = 1; //cells in memory
    
    String bOutput = ""; //binary output
    String cOutput = ""; //character output
    char c; //code command
    
    Scanner input = new Scanner(System.in);
    String f = ""; //for functions
    
    //temporary variables
    String t;
    int x;
    
    //------------------------------------------------------------------------
    
    //inserts function code when function call
    //pre-execution
    while (i < l) {
      c = code.charAt(i);
      switch (c) {
        //declare function  
        case 'D':
          i++;
          f = "";
          while (code.charAt(i) != 'D') {
            f += code.charAt(i);
            i++;
          }
          break;
        //enter function into code 
        case 'E':
          t = "";
          for (x = 0; x < i+1; x++) {
            t += code.charAt(x);
          }
          t += f;
          for (x = i+1; x < l; x++) {
            t += code.charAt(x);
          }
          code = t;
          l = code.length();
          break;
      }
      i++;
    }
    i = 0;
    
    //--------------------------------------------------------------
    
    //reads and executes the code
    while (i < l) {
      c = code.charAt(i);

      switch (c) {
        
        //pointer movement
        case '0':
          ptr = 0;
          break;
        case '1':
          ptr++;
          if (ptr == cells) {
            memory.add((byte) 0);
            cells++;
          }
          break;
        case '2':
          ptr--;
          if (ptr < 0) {
            ptr = cells - 1;
          } 
          break;
        
        //bit flip
        case '3':
          if (memory.get(ptr) == 1) {
            memory.set(ptr,(byte) 0);
          }
          else {
            memory.set(ptr,(byte) 1);
          }
          break;

        case '4':
          bOutput += memory.get(ptr);
          break;
          
        //if-else-endif  
        case '5':
          if (memory.get(ptr) == 0) {
            x = 1;
            while (x != 0) {
              i++;
              if (code.charAt(i) == '5') {
                x++;
              }
              else if (code.charAt(i) == '7' || (x == 1 && code.charAt(i) == '6')) {
                x--;
              }
            }
          }
          break;
        //else
        case '6':
          x = 1;
          while (x != 0) {
            i++;
            if (code.charAt(i) == '5') {
              x++;
            }
            else if (code.charAt(i) == '7') {
              x--;
            }
          }
          break;
          
        case '7':
          //endif command, does nothing on its own
          break;
          
        //loop  
        case '8':
          if (memory.get(ptr) == 1) {
            x = 1;
            i--;
            while (x != 0) {
              i--;
              if (code.charAt(i) == '7') {
                x++;
              }
              else if (code.charAt(i) == '5') {
                x--;
              }
            }
          }
          break;
        
        case '9':
          memory.set(ptr, (byte) Math.round(Math.random()));

        //input
        case 'A':
          x = Integer.parseInt(input.nextLine());
          if (x != 0 && x != 1) {
            System.out.println(1111);
            i = l;
          }
          else {
            memory.set(ptr,(byte) x);
            break;
          }
        
        //binary output
        case 'B':
          System.out.println(bOutput);
          bOutput = "";
          break;
       
        //integer and character output
        case 'C':
          if (bOutput.equals("")) {
            System.out.println(cOutput);
            cOutput = "";
          }
          else if (memory.get(ptr) == 1) {
            cOutput += toDec(bOutput);
          }
          else {
            cOutput += (char) toDec(bOutput);
          }
          bOutput = "";
          break;
          
        case 'D':
          break;
        case 'E':
          //D and E are handled pre-execution
          break;
          
        case 'F':
          System.out.println(1111);
          i = l;
      }
     
      i++;
    }  
  input.close();
  }
}

File: acl.py
#Advanced Computer Language --> Python port

import random

class ACL:
    def getCode(self) -> str:
        try:
            print("File to interpret? ")
            fileName: str = input()

            file = open(fileName)
            code: str = file.read()
            file.close()

            return code
        except FileNotFoundError as e:
            print("FILE FAIL")
            return "F"

    def toDec(self, _bin: str) -> int:
        l: int = len(_bin)
        j: int = l - 1
        x: int = 0
        dec: int = 0
        while x < l:
            dec += (2 ** x) * (ord(_bin[j]) - 48)
            x += 1; j -= 1

        return dec

    def main(self):

        #declares necessary variables etc.
        
        code: str = self.getCode()
        l: int = len(code)
        i: int = 0 #code reader pointer

        memory: list(int) = list() #cell based, binary
        memory.append(int(0)) #make the memory have one cell at start
        ptr: int = 0 #memory pointer
        cells: int = 1 #cells in memory

        bOutput: str = "" #binary output
        cOutput: str = "" #character output
        c: str = "" #code command

        f: str = "" #for functions

        #temporary variables
        t: str = ""
        x: int = 0


        #--------------------

        #inserts function code when function call
        #pre-execution
        while i < l:
            c = code[i]
            #declare function
            if c == "D":
                i += 1
                f = ""
                while code[i] != "D":
                    f += code[i]
                    i += 1
            #enter function into code
            elif c == "E":
                t = ""
                for x in range(0, i + 1):
                    t += code[x]
                t += f

                for x in range(i + 1, l):
                    t += code[x]

                code = t
                l = len(code)

            i += 1

        i = 0

        #--------------------

        while i < l:
            c = code[i]

            #pointer movement
            if c == "0":
                ptr = 0

            elif c == "1":
                ptr += 1
                if ptr == cells:
                    memory.append(int(0))
                    cells += 1


            elif c == "2":
                ptr -= 1
                if ptr < 0:
                    ptr = cells - 1

            #bit flip
            elif c == "3":
                if memory[ptr] == 1:
                    memory[ptr] = int(0)
                else:
                    memory[ptr] = int(1)

            elif c == "4":
                bOutput += str(memory[ptr])

            #if-else-endif
            elif c == "5":
                if memory[ptr] == 0:
                    x = 1
                    while x != 0:
                        i += 1
                        if code[i] == '5':
                            x += 1
                        elif code[i] == "7" or (x == 1 and code[i] == "6"):
                            x -= 1

            #else
            elif c == "6":
                x = 1
                while x != 0:
                    i += 1
                    if code[i] == "5":
                        x += 1
                    elif code[i] == "7":
                        x -= 1

            elif c == "7":
                #endif command, does nothing on its own
                pass

            #loop
            elif c == "8":
                if memory[ptr] == 1:
                    x = 1
                    i -= 1
                    while x != 0:
                        i -= 1
                        
                        if code[i] == "7":
                            x += 1
                        elif code[i] == "5":
                            x -= 1

            elif c == "9":
                memory[ptr] = int(round(random.random()))

            #input
            elif c == "A":
                x = int(input())
                if (x != 0 and x != 1):
                    print(1111)
                    i = l

                else:
                    memory[ptr] = int(x)

            #binary output
            elif c == "B":
                print(bOutput)
                bOutput = ""

            #integer and character output
            elif c == "C":
                if bOutput == "":
                    print(cOutput)
                    cOutput = ""
                elif memory[ptr] == 1:
                    cOutput += str(self.toDec(bOutput))
                else:
                    cOutput += str(chr(self.toDec(bOutput)))

                bOutput = ""

            elif c == "D":
                pass

            elif c == "E":
                #D and E are handled pre-execution
                pass

            elif c == "F":
                print(1111)
                i = l

            i += 1


interpreter = ACL()
interpreter.main()

